mattermost: 9.11.11 -> 10.5.2; use formats.json; squash systemd-tmpfiles warning

numinit opened the pull request:
There was a systemd-tmpfiles warning about not being able to remove the 'plugins' directory. Squash this warning through removal of unnecessary systemd-tmpfiles options, and write a test for it.

Add the {option}`services.mattermost.pluginsBundle` option to allow overriding the plugin directory and also using it for tests. Update wording in option documentation so it is more clear.

Use formats.json instead of builtins.toJSON so module merging works.

Make the tests go faster by pipelining shutdowns of nodes.

Followup to #208181 which introduced these minor nits.


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

numinit committed:
mattermost: squash systemd-tmpfiles warning (+ other minor fixes)

There was a systemd-tmpfiles warning about not being able to remove the
'plugins' directory. Squash this warning through removal of unnecessary
systemd-tmpfiles options, and write a test for it.

Add the {option}`services.mattermost.pluginsBundle` option to allow
overriding the plugin directory and also using it for tests. Update
wording in option documentation so it is more clear.

Use formats.json instead of builtins.toJSON so module merging works.

Make the tests go faster by pipelining shutdowns of nodes.

numinit commented on nixos/modules/services/web-apps/mattermost.nix:
```diff
@@ -756,7 +767,7 @@ in
           "d= ${tempDir} 0750 ${cfg.user} ${cfg.group} - -"
 
           # Ensure that pluginDir is a directory, as it could be a symlink on prior versions.
-          "r- ${pluginDir} - - - - -"
```
These `r-` caused the warning. `d=` will make sure they're a directory instead of a symlink to one.

numinit commented on nixos/modules/services/web-apps/mattermost.nix:
```diff
@@ -147,32 +147,29 @@ let
-    with pkgs;
-    map (
-      plugin:
-      stdenv.mkDerivation {
```
No CC needed in these, and we can symlink prior to repacking.

numinit commented on nixos/tests/mattermost/default.nix:
```diff
@@ -342,39 +344,54 @@ import ../make-test-python.nix (
-                                 mostlyMutableToplevel: str, mostlyMutable,
+        def switch_to_specialisation(node, toplevel: str, specialisation: str):
+          print(f"switch_to_specialisation({node.name!r}, {toplevel!r}, {specialisation!r})", file=sys.stderr)
+          node.succeed(f"{toplevel}/specialisation/{specialisation}/bin/switch-to-configuration switch || true")
```
For a variety of reasons, the config switch can fail for unrelated reasons to Mattermost if we're on unstable, but Mattermost may still restart. We should just wait for it to come up.

numinit commented:
Rebased against master and added formats support.

drupol commented on nixos/modules/services/web-apps/mattermost.nix:
```diff
@@ -147,32 +147,29 @@ let
+    plugin:
+    pkgs.stdenvNoCC.mkDerivation {
+      name = "${cfg.package.name}-plugin";
+      installPhase = ''
```
Missing phase hooks.

numinit commented:
Should be good to go.

RossComputerGuy reviewed (approved):
Diff LGTM, the tests should build fine for me but we'll see.

RossComputerGuy commented:
![image](https://github.com/user-attachments/assets/4d58853b-b254-4aff-9f74-bddd899261d0)
Tests are still running

numinit commented:
Rebased, let's try that.

RossComputerGuy commented:
Alright, trying it now.

numinit commented:
Changed nixos/mattermost -> mattermost in the commit message so eval works.

numinit commented:
Seems to fail on aarch64 with:

--- FAIL: TestUploadFiles/simple_Happy_image_thumbnail/preview_6_[orientation_test_6.jpeg] (1.76s)
FAIL channels/api4.TestUploadFiles/simple_Happy_image_thumbnail/preview_6_[orientation_test_6.jpeg] (re-run 1) (1.76s)

So the thumbnail generation for posts is nondeterministic on aarch64...

numinit commented:
Same failure as https://github.com/NixOS/nixpkgs/pull/389563/checks?check_run_id=38752816065

numinit committed:
mattermost: disable failing aarch64 tests

These were added fairly recently, and apparently test that thumbnailing
an image exactly matches a reference image, which varies by arch.

numinit commented:
@ofborg build mattermost

numinit commented:
@ofborg build nixosTests.mattermost

numinit committed:
nixos/mattermost: increase db connect timeouts

numinit commented:
@ofborg build nixosTests.mattermost

numinit commented:
Strange. I get successful nixpkgs-review builds on the community aarch64 builder and my personal EPYC 4 machine, but a failed ofborg build involving MariaDB.

numinit committed:
nixos/tests/mattermost: disable MySQL tests on aarch64

There seems to be some intermittent issue initializing the database on
aarch64. Still test Postgres since it is the default.

numinit committed:
mattermost: build without long test suite by default

It was good for the unstable cycle to sort things out, switch it for
25.05 release since it's still in the passthru.

numinit committed:
mattermost: 9.11.11 -> 10.5.2

numinit commented:
Just disabled the MySQL tests on aarch64. It's already slow enough on x86_64, and Postgres is the default anyway.

numinit commented:
`nix-build -A nixosTests.mattermost` looks good and `nix-build -A mattermost.withTests` is currently finishing. Made the huge test suite not the default derivation as we stabilize unstable. It was useful for a while. :-)

RossComputerGuy commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 376838`

---
### `aarch64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>gclient2nix</li>
    <li>mattermost</li>
    <li>mattermostLatest</li>
    <li>mmctl</li>
  </ul>
</details>
